en: stop tidb gracefully (#611)
pingcap/docs-tidb-operator: en/access-tidb.md

- `spec.tidb.terminationGracePeriodSeconds`: The longest duration to delete the old TiDB pod during the rolling upgrade. If this duration is exceeded, the TiDB pod will be deleted forcefully.
- `spec.tidb.lifecycle`: Sets the `preStop` hook for the TiDB pod, which is the operation executed before the TiDB server stops.

```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic
spec:
  version: v4.0.0
  pvReclaimPolicy: Retain
  discovery: {}
  pd:
    baseImage: pingcap/pd
    replicas: 1
    requests:
      storage: "1Gi"
    config: {}
  tikv:
    baseImage: pingcap/tikv
    replicas: 1
    requests:
      storage: "1Gi"
    config: {}
  tidb:
    baseImage: pingcap/tidb
    replicas: 1
    service:
      type: ClusterIP
    config: {}
    terminationGracePeriodSeconds: 60
    lifecycle:
      preStop:
        exec:
          command:
          - /bin/sh
          - -c
          - "sleep 10 && kill -QUIT 1"
```

The YAML file above:

- Sets the longest duration to delete the TiDB pod to 60 seconds. If the client does not close the connections after 60 seconds, these connections will be closed forcefully. You can adjust the value according to your needs.

> - Sets the `preStop` hook to `sleep 10 && kill -QUIT 1`. Here Pid 1 refers to the Pid of the TiDB server process in the TiDB pod. When the TiDB server process receives the signal, it exits only after all the connections are closed by the client.

Reviewer: ```suggestion
- Sets the value of `preStop` hook to `sleep 10 && kill -QUIT 1`. Here `PID 1` refers to the PID of the TiDB server process in the TiDB Pod. When the TiDB server process receives the signal, it exits only after all the connections are closed by the client.
```

> When Kubernetes deletes the TiDB pod, it also removes the TiDB node from the service endpoints. This is to ensure that the new connection does not connect to this TiDB node. However, since this process is asynchronous, you can make the system sleep for a few seconds before you send the kill signal, which makes sure that the TiDB node is removed from the endpoints.

Reviewer: ```suggestion
When Kubernetes deletes the TiDB Pod, it also removes the TiDB node from the service endpoints. This is to ensure that the new connection is not established to this TiDB node. However, because this process is asynchronous, you can make the system sleep for a few seconds before you send the `kill` signal, which makes sure that the TiDB node is removed from the endpoints.
```
